### PR TITLE
bugfixes to reportEmployment()

### PR DESCRIPTION
**Pull request by amnmalik**

### CODE REVIEW

**Review by pfuehrlich-pik — APPROVED** [review]

**Comment by pfuehrlich-pik** [comment]
You need to rerun buildLibrary, I think you version increment got lost during a merge.

**Comment by amnmalik** [comment]
@pfuehrlich-pik I don't know why I am seeing lucode2 checks failing now. I thought this pull-request was disposed. Do you know what happened?

File: R/reportEmployment.R [hunk 1]
```diff
@@ -11,15 +10,16 @@
...
+#' reportEmployment(gdx, improvements = "All", multiplier = "own", subtype = "expert", share_manf = "local", decline = "capcosts")
 #' }
 #' @importFrom madrat calcOutput
 #' @importFrom magclass getNames add_columns getItems
 #' @export
 
-reportEmployment <- function(gdx, improvements, multiplier, subtype, share_manf, decline, type) {
+reportEmployment <- function(gdx, improvements, multiplier, subtype, share_manf, decline) {
 
+  # use setConfig(forcecache=T) to make the code run faster
```
Comment (pfuehrlich-pik) [inline R/reportEmployment.R:22]: This seems like a dangerous hint as it is. I suggest either explaining a bit further in which situation forcecache might be helpful and what problems might arise from it, or to delete this comment and leave explaining forcecache to the madrat package.

File: R/runEmployment.R [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+#' @export
+
+reportEmployment <- function(pathToMIF, improvements, multiplier, subtype, share_manf, decline) {
+  
+  input_mif <- remind2::readReportingMIF(pathToMIF = "") # read as data frame
+  input_mif_mp <- as.magpie(input_mif) # convert to magpie object
+  input_mif_mp <- collapseDim(input_mif_mp) # remove dimensions with same values, i.e., "model" and "scenario"
+  input_mif_mp <- collapseDim(input_mif_mp,dim = 3.2) # remove "unit" dimension
+  
+  # use setConfig(forcecache=TRUE) to make the code run faster
```
Comment (pfuehrlich-pik) [inline R/runEmployment.R:27]: See my comment above

File: R/runEmployment.R [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+  getNames(x) <- gsub("Wind offshore","Wind|Offshore",getNames(x))
+  x <- x[, , "HP", pmatch = TRUE, invert = TRUE] # excluding (at the moment) jobs in combined heat and power plants
+  
+  ## DECLINE FACTORS------------------------
+  if (decline == "capcosts") {
+    
+
+    # capital costs and OM fixed costs evolution over time for different techs, used to calculated the decline factor
+   # report_tech <- reportTechnology(gdx)
+    #report_tech <- collapseNames(report_tech)
```
Comment (pfuehrlich-pik) [inline R/runEmployment.R:40]: Can you explain in a comment why these lines are commented out and when/under which circumstances they should be commented in again? If not, maybe it's better to delete them for now?

File: R/runEmployment.R [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+             "Tech|Electricity|Wind|Offshore|OM Cost|fixed",
+             "Tech|Electricity|Biomass|Combined Heat and Power w/o CC|OM Cost|fixed",
+             "Tech|Electricity|Oil|DOT|OM Cost|fixed",
+             "Tech|Electricity|Geothermal|OM Cost|fixed",
+             "Tech|Electricity|Storage|Battery|For PV|OM Cost|fixed", 
+             NULL
+    )
+    # only capital costs
+    report_tech <- input_mif_mp[,,var]
+    cap_costs <- report_tech[getRegions(x), seq(2015, 2050, 5), "Capital Costs",pmatch=TRUE]# selecting years until 2050 and removing "World" region
```
Comment (pfuehrlich-pik) [inline R/runEmployment.R:72]:
```suggestion
    # selecting years until 2050 and removing "World" region
    cap_costs <- report_tech[getRegions(x), seq(2015, 2050, 5), "Capital Costs", pmatch = TRUE]
```

File: R/runEmployment.R [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+    
+    # Evolution of employment factor (CI & Manf) with time depends on capital costs
+    # the job intensity decreases with capital costs for techs, relative to 2015
+    var_in_x <- getItems(x, dim = 3.1) # variables from employment factor magpie object
+    var_in_rem <- c("Coal", "Gas", "Nuclear", "Biomass", "Hydro", "Hydro", "Wind|Onshore", "Wind|Offshore", "Solar|PV",
+                    "Geothermal", "Solar|CSP", "Oil", "Solar|PV", "Storage") # variables from remind (order matters)
+    var_comb <- paste(var_in_x, var_in_rem, sep = ".") # concatenating the two for looping to work
+    
+    for (i in var_comb) {
+      x[getRegions(x), , sub("\\..*", "", i)][, , c("CI", "Manf")] <- x[getRegions(x), , sub("\\..*", "", i)][, , c("CI", "Manf")] * setNames(cap_costs[getRegions(x), getYears(x), sub(".*\\.", "", i), pmatch = TRUE], NULL)
```
Comment (pfuehrlich-pik) [inline R/runEmployment.R:84]: Please try to split up lines that are longer than 120 characters into smaller simpler steps to improve readability. If that is not possible please add line breaks.

File: R/runEmployment.R [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+  ## Other external parameters--------------------------
+
+  # share of rooftop in total spv, wind offshore in total wind, and hydro small in total hydro
+  share <- calcOutput("DspvShare")
+  
+  ## ----------------------------------------------
+  
+  ## New and Absolute capacity from REMIND---------------------------------
+  # filtering required capacity variables i.e., only new and existing coal capacities
+#  remind_output <- reportCapacity(gdx)
```
Comment (pfuehrlich-pik) [inline R/runEmployment.R:116]: Can you explain in a comment why these lines are commented out and when/under which circumstances they should be commented in again? If not, maybe it's better to delete them for now?

File: R/runEmployment.R [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+    # 2015 and 2020 values same as 2019 values
+    prod_share_tmp[, c(2015, 2020), ] <- prod_share[, , getNames(prod_share_tmp)]
+    # from 2020 to 2050, linearly interpolate
+    prod_share <- time_interpolate(prod_share_tmp, seq(2025, 2045, 5), integrate_interpolated_years = TRUE)
+    if (share_manf == "current") {
+      prod_share[, , ] <- prod_share[, 2020, ] # shares remain same throughout = 2020
+    }
+  }
+  # Obtaining fuel supply variables
+#  remind_prod <- reportExtraction(gdx)["GLO", , , invert = TRUE] # remove GLO
```
Comment (pfuehrlich-pik) [inline R/runEmployment.R:203]: Please see my previous comments on commented out code.

File: R/runEmployment.R [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+  remind_prod <- input_mif_mp[getRegions(x), getYears(x), vars] # period >2010 and only select variables
+  
+  pe_trad <- input_mif_mp[getRegions(x), getYears(x), "PE|Biomass|Traditional"]
+  
+  # removing the traditional biomass component, where biomass isn't sold, thus no employment
+  # in conventional sense
+  remind_prod[, , "PE|Production|Biomass"] <-  remind_prod[, , "PE|Production|Biomass"] - setNames(pe_trad[, , ], NULL)
+  
+  # getNames(remind_prod) <- gsub(pattern = "\\[Energy\\]",replacement = "",x = getNames(remind_prod))
+ # getNames(remind_prod) <- gsub(pattern = "  \\(EJ\\/yr\\)", replacement = " \\(EJ\\/yr\\)", x = getNames(remind_prod)) # removing space
```
Comment (pfuehrlich-pik) [inline R/runEmployment.R:219]: Please see my previous comments on commented out code.

File: R/runEmployment.R [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+  getNames(remind_prod) <- gsub(pattern = "Uranium \\[Energy\\]", replacement = "Nuclear", x = getNames(remind_prod)) # removing space
+  
+  ## Calculating jobs--------------------------------
+  techs <- c("Solar|PV-utility", "Solar|PV-rooftop", "Hydro-large", "Hydro-small",
+             "Wind|Offshore", "Wind|Onshore", "Coal", "Biomass", "Gas", "Storage", "Oil", "Nuclear", "Geothermal", "Solar|CSP")
+  techs_exp <-  c("Solar|PV-utility", "Solar|PV-rooftop", "Wind|Offshore", "Wind|Onshore") # technologies with export component
+  
+  # cons <- new.magpie(getRegions(x),names = techs)
+  # average construction time for different technologies
+  # cons[] <- rep(c(1,1,4,3,4,2,5,2,2,1,2,10,2,2),each=12)
```
Comment (pfuehrlich-pik) [inline R/runEmployment.R:229]: Please see my previous comments on commented out code.

File: R/runEmployment.R [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+  
+  # 1. Manufacturing jobs for techs with export component.
+  
+  # 1a) Total manf jobs (per region) for these techs are equal to total world addition X share of
+  # world exports (for that region)
+  
+  jobs_manf <- new.magpie(getRegions(x), getYears(x), techs, fill = 0)
+  for (i in techs_exp) {
+    # manf <- paste0(i,".","Manf")
```
Comment (pfuehrlich-pik) [inline R/runEmployment.R:238]: Please see my previous comments on commented out code.

File: R/runEmployment.R [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+    jobs_manf[, , i] <- added_cap_sum[, getYears(x), i, pmatch = TRUE] * 1000
+    jobs_manf[, getYears(x), i] <-  (jobs_manf[, getYears(x), i] * setNames(x[, , i][, , "Manf"], NULL) * prod_share[, , i])
+  }
+  
+  for (i in setdiff(techs, techs_exp)) {
+    # manf <- paste0(i,".","Manf")
```
Comment (pfuehrlich-pik) [inline R/runEmployment.R:244]: Please see my previous comments on commented out code.

File: R/runEmployment.R [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+  getSets(jobs_manf) <- c("region", "year", "variable", "type")
+  
+  # 2. Construction and Installation jobs
+  jobs_ci <- new.magpie(getRegions(x), getYears(x), techs, fill = 0)
+  
+  
+  for (i in techs) {
+    # inst <- paste0(i,".","CI")
+    jobs_ci[, , i] <- added_cap[, getYears(x), i, pmatch = TRUE] * 1000
+    # jobs_ci[,,i] <-  (jobs_ci[,getYears(x),i] *  setNames(x[,,i,pmatch=TRUE][,,"CI"],NULL))
```
Comment (pfuehrlich-pik) [inline R/runEmployment.R:259]: Please see my previous comments on commented out code.

File: R/runEmployment.R [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+  
+  jobs_ci <- add_dimension(jobs_ci, dim = 3.2, add = "type", nm = "CI")
+  getSets(jobs_ci) <- c("region", "year", "variable", "type")
+  
+  
+  ## 3. Jobs in O & M
+  # jobs = existing capacity per tech * emp factor per tech
+  jobs_om <- new.magpie(getRegions(x), getYears(x), techs)
+  for (i in techs) {
+    # inst <- paste0(i,".","OM")
```
Comment (pfuehrlich-pik) [inline R/runEmployment.R:272]: Please see my previous comments on commented out code.

File: R/runEmployment.R [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+  }
+  jobs_om <- add_dimension(jobs_om, dim = 3.2, add = "type", nm = "OM")
+  getSets(jobs_om) <- c("region", "year", "variable", "type")
+  
+  ## 4. Jobs in Fuel supply
+  # jobs = total production (in EJ/yr) per fuel * employment factor per fuel (Jobs/PJ)
+  fuels <- c("Coal", "Gas", "Biomass", "Oil")
+  jobs_prod <- new.magpie(getRegions(x), getYears(x), techs, fill = 0)
+  for (i in fuels) {
+    # fs <- paste0(i,".","Fuel_supply")
```
Comment (pfuehrlich-pik) [inline R/runEmployment.R:284]: Please see my previous comments on commented out code.

File: R/runEmployment.R [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+    jobs_prod[, , i] <- remind_prod[, getYears(x), i, pmatch = TRUE] * 1000
+    jobs_prod[, , i] <-  jobs_prod[, , i] * setNames(x[, , i][, , "Fuel_supply"], NULL)
+  }
+  jobs_prod <- add_dimension(jobs_prod, dim = 3.2, add = "type", nm = "Fuel_supply")
+  getSets(jobs_prod) <- c("region", "year", "variable", "type")
+  
+  ## Jobs in Fuel supply, nuclear which are given in terms of SE
+#  rem_se_nuc <- reportSE(gdx)[regions, getYears(x), "Nuclear", pmatch = TRUE] * 277777.778 # EJ to GWh
```
Comment (pfuehrlich-pik) [inline R/runEmployment.R:292]: Please see my previous comments on commented out code.

**Review by pfuehrlich-pik — COMMENTED** [review]
Thanks for your contribution! I believe the readability of your code could be improved by running `lucode2::autoFormat()` and fixing linter warnings. To help with converting your snake_case variable names to camelCase you can use `lucode2::snakeToCamel(<path>)`.

File: R/reportEmployment.R [hunk 16]
```diff
@@ -11,15 +10,16 @@
 #' @author Aman Malik
 #' @examples
 #' \dontrun{
-#' reportEmployment(gdx, improvements = "All", multiplier = "own", subtype = "expert", 
-#' share_manf = "local", decline = "capcosts",type="per_year")
+
+#' reportEmployment(gdx, improvements = "All", multiplier = "own", subtype = "expert", share_manf = "local", decline = "capcosts")
```
Comment (pfuehrlich-pik) [inline R/reportEmployment.R:14]:
```suggestion
#' reportEmployment(gdx, improvements = "All", multiplier = "own", subtype = "expert",
#'                  share_manf = "local", decline = "capcosts")
```

Comment (pfuehrlich-pik) [reply R/reportEmployment.R:14]: Examples must not have lines with more than 100 characters, or they will get a NOTE which prevents buildLibrary from successfully finishing.

File: R/runEmployment.R [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+#' @param share_manf Either "current" or "local". Current implies current shares of world manufacture remain same until 2050, current means that in 2050 all countries manufacture required components locally.
+#' @param multiplier controls how the regional multiplier for non-oecd countries changes with time.
+#' @param decline How should the employment factors change over time? "capcosts" means according to capital costs. "static" means it doesn't change
+#' @description This function returns a magpie object containing the reporting the jobs for different technologies
+#' @return A magpie object
+#' @author Aman Malik
+#' @examples
+#' \dontrun{
+
+#' runEmployment(pathToMIF, improvements = "All", multiplier = "own", subtype = "expert", share_manf = "local", decline = "capcosts")
```
Comment (pfuehrlich-pik) [inline R/runEmployment.R:14]: Examples must not have lines with more than 100 characters, or they will get a NOTE which prevents buildLibrary from successfully finishing.
```suggestion
#' runEmployment(pathToMIF, improvements = "All", multiplier = "own", subtype = "expert",
#'               share_manf = "local", decline = "capcosts")
```

File: R/reportEmployment.R [hunk 18]
```diff
@@ -1,6 +1,5 @@
 #' Computes the employment values (jobs) across different sectors
 #' @param gdx A gdx file output from a REMIND run
```
Comment (pfuehrlich-pik) [inline R/reportEmployment.R:2]: There's a warning in the [GitHub action check](https://github.com/pik-piam/remind2/pull/104/checks) saying that `gdx` is not documented, but here it is. Did you run `lucode2::buildLibrary()`? That should among other things build the documentation

**Review by pfuehrlich-pik — CHANGES_REQUESTED** [review]
Please make sure that `lucode2::buildLibrary()` runs through successfully and increase the package version number.

File: R/runEmployment.R (hunk 17 above)
Comment (amnmalik) [reply R/runEmployment.R:14]: 100 or 120? lucode tells me 120 characters.

File: R/runEmployment.R (hunk 5 above)
Comment (amnmalik) [reply R/runEmployment.R]: Did where ever possible without affecting code readability

File: R/runEmployment.R (hunk 11 above)
Comment (amnmalik) [reply R/runEmployment.R]: removed commented code

File: R/runEmployment.R (hunk 12 above)
Comment (amnmalik) [reply R/runEmployment.R]: removed commented code

File: R/runEmployment.R (hunk 13 above)
Comment (amnmalik) [reply R/runEmployment.R]: removed commented code

File: R/reportEmployment.R (hunk 18 above)
Comment (amnmalik) [reply R/reportEmployment.R:2]: works after new commits

File: R/runEmployment.R (hunk 17 above)
Comment (pfuehrlich-pik) [reply R/runEmployment.R]: In code it's 120 which is how we configured our linter. For examples the standard R CMD CHECK defines the limit as 100.

**Pull request merged**